src/libstd/sys/windows/fs.rs

```diff
@@ -561,37 +617,148 @@ pub fn rmdir(p: &Path) -> io::Result<()> {
+        remove_item(path.as_ref(), &mut ctx)
     } else {
-        remove_dir_all_recursive(path)
+        Err(io::Error::new(io::ErrorKind::PermissionDenied, "Not a directory"))
     }
 }
 
-fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
+struct RmdirContext<'a> {
+    base_dir: &'a Path,
+    readonly: bool,
+    counter: u64,
+}
+
+fn remove_dir_all_recursive(path: &Path, ctx: &mut RmdirContext)
+                            -> io::Result<()> {
+    let dir_readonly = ctx.readonly;
     for child in try!(readdir(path)) {
         let child = try!(child);
         let child_type = try!(child.file_type());
+        ctx.readonly = try!(child.metadata()).perm().readonly();
         if child_type.is_dir() {
-            try!(remove_dir_all_recursive(&child.path()));
-        } else if child_type.is_symlink_dir() {
-            try!(rmdir(&child.path()));
+            try!(remove_dir_all_recursive(&child.path(), ctx));
         } else {
-            try!(unlink(&child.path()));
+            try!(remove_item(&child.path().as_ref(), ctx));
         }
     }
-    rmdir(path)
+    ctx.readonly = dir_readonly;
+    remove_item(path, ctx)
+}
+
+fn remove_item(path: &Path, ctx: &mut RmdirContext) -> io::Result<()> {
+    if !ctx.readonly {
+        let mut opts = OpenOptions::new();
+        opts.access_mode(c::DELETE);
+        opts.custom_flags(c::FILE_FLAG_BACKUP_SEMANTICS | // delete directory
+                          c::FILE_FLAG_OPEN_REPARSE_POINT | // delete symlink
+                          c::FILE_FLAG_DELETE_ON_CLOSE);
+        let file = try!(File::open(path, &opts));
+        move_item(&file, ctx)
+    } else {
+        // remove read-only permision
+        try!(set_perm(&path, FilePermissions::new()));
+        // move and delete file, similar to !readonly.
+        // only the access mode is different.
+        let mut opts = OpenOptions::new();
+        opts.access_mode(c::DELETE | c::FILE_WRITE_ATTRIBUTES);
+        opts.custom_flags(c::FILE_FLAG_BACKUP_SEMANTICS |
+                          c::FILE_FLAG_OPEN_REPARSE_POINT |
+                          c::FILE_FLAG_DELETE_ON_CLOSE);
+        let file = try!(File::open(path, &opts));
+        try!(move_item(&file, ctx));
+        // restore read-only flag just in case there are other hard links
+        let mut perm = FilePermissions::new();
+        perm.set_readonly(true);
+        let _ = file.set_perm(perm); // ignore if this fails
+        Ok(())
+    }
+}
+
+fn move_item(file: &File, ctx: &mut RmdirContext) -> io::Result<()> {
+    let mut tmpname = ctx.base_dir.join(format!{"rm-{}", ctx.counter});
+    ctx.counter += 1;
+    // Try to rename the file. If it already exists, just retry with an other
+    // filename.
+    while let Err(err) = file.rename(tmpname.as_ref(), false) {
+        if err.kind() != io::ErrorKind::AlreadyExists { return Err(err) };
+        tmpname = ctx.base_dir.join(format!("rm-{}", ctx.counter));
+        ctx.counter += 1;
+    }
```

Comment:
Could the `join` + increment be deduplicated here with something like:

```rust
loop {
    let tmpname = foo.join(bar);
    counter++;
    match rename() {
        Ok(()) => return Ok(()),
        Err(ref e) if e.is_already_exists() => {}
        Err(e) => return Err(e),
    }
}
```

It may also be a good idea to have the basename be something obscure like `__rust__rm__` and to also cap the number of iterations done here. It looks like a bunch of concurrent `remove_dir_all` could cause a lot of contention on the filesystem here perhaps?